### PR TITLE
Geometry types

### PR DESCRIPTION
**Pull request by jeroenbourgois**
Hip Hop Hurray! Thanks to awesome input from @wojtekmach and some plumbing on beforehand from our part, we landed a good basis for geo support in `MyXQL`! Already in use in production!

Hope this can get into master soon, some other folks are waiting on it, and then it can get tested some more.

Regards
Jack + Joe (and @wojtekmach !)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -126,6 +126,57 @@ You can customize it to use another library via the `:json_library` configuratio
...
+        properties: %{},
+        srid: nil
+      }
+    ]
+  ]
+}
+```
+
+If you plan on using MyXQL with Ecto, and need to handle user input, you might need to add a [custom Ecto type](https://hexdocs.pm/ecto/Ecto.Type.html
+) in order to construct a valid WKT string.
```
Comment (wojtekmach) [inline README.md]: this is a great start!

> `MyXQL.query!(pid, "INSERT INTO geo (`point`) VALUES (ST_GeomFromText('#{Geo.WKT.encode!(point)}'))")`

by default the queries we execute are prepared which means we use the binary wire protocol and because we already encode the values, this would be shorter and more efficient :)

```elixir
MyXQL.query!(pid, "INSERT INTO geo (`point`) VALUES (?)", [point])
```

there's couple more minor changes that I'd like to make and I'm just gonna go ahead and change things out a bit on this branch and I'd appreciate feedback!

File: lib/myxql/protocol/values.ex [hunk 2]
```diff
@@ -381,29 +392,23 @@ defmodule MyXQL.Protocol.Values do
...
-  #
-  # so yeah, looks like mysql might be storing it in an internal format that is not WKB.
-  # There's a ST_ToBinary() function that allegedly returns data in WKB so this might be helpful
-  # for debugging.
-  #
-  # [1] https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry#Well-known_binary
-  # [2] https://dev.mysql.com/doc/refman/8.0/en/populating-spatial-columns.html
+  # Still, all considered, this is a very elegant and readable solution!
   defp decode_geometry(<<0::uint4, r::bits>>) do
     r |> Base.encode16() |> Geo.WKB.decode!()
```
Comment (wojtekmach) [inline lib/myxql/protocol/values.ex]: for the record this is not the most efficient way to handle this as we're doing the intermediate conversion to base16 before decoding it again. I added an issue https://github.com/bryanjos/geo/issues/121 to improve things, I'm done with the easy part (encoding: https://gist.github.com/wojtekmach/64a49ccb2948cb0a5009e97a11138a50) and still need to do decoding, but if someone wants to beat me to it, by all means!

File: .gitignore [hunk 3]
```diff
@@ -25,3 +25,4 @@ myxql-*.tar
 /errmsg-utf8.txt
 /xref_graph.*
 /config
+.elixir_ls
```
Comment (wojtekmach) [inline .gitignore]: could you revert this change? We try to keep such editor-specific changes out and instead ask folks to make them in their respective ~/.gitignore_global.

**Comment by wojtekmach** [comment]
Forgot to mention...

> Already in use in production!

That's fantastic news!

File: README.md (hunk 1 above)
Comment (wojtekmach) [reply README.md]: here's the changes I had in mind: https://gist.github.com/wojtekmach/8643f1aa2ef30fd81f393c17137a89cd

What do you think?

My rationale for keeping it short is people that will use the `POLYGON` type or the `ST_GeomFromText` function will likely already know them from reading the MySQL user guide - and if we remove this information, what's left is general usage of MyXQL that we explain earlier.

Comment (wojtekmach) [reply README.md]: I went ahead and edited it on your branch. Feedback still welcome though!

**Comment by wojtekmach** [comment]
@jeroenbourgois oh, I didn't notice you've opened this branch against mine! If it's not too much trouble, would you mind changing base branch to `master`? (or just opening a new PR if that's easier)

Oh, and if you want to do the honours and remove https://github.com/elixir-ecto/myxql/blame/master/MARIAEX_COMPATIBILITY.md#L41 - by all means :D Thanks again for your help with this!

**Comment by jeroenbourgois** [comment]
@wojtekmach ok, will do, somewhere tonight probably.

File: .gitignore (hunk 3 above)
Comment (jeroenbourgois) [reply .gitignore]: Done

File: lib/myxql/protocol/values.ex (hunk 2 above)
Comment (jeroenbourgois) [reply lib/myxql/protocol/values.ex]: Ok!

File: README.md (hunk 1 above)
Comment (jeroenbourgois) [reply README.md]: Looks good

File: test/test_helper.exs [hunk 4]
```diff
@@ -126,7 +126,8 @@ defmodule TestHelper do
       #{if supports_json?(), do: "my_json JSON,", else: ""}
       my_char CHAR,
       my_point POINT,
-      my_multipoint MULTIPOINT
+      my_multipoint MULTIPOINT,
+      my_polygon POLYGON
```
[the remark this answers is not in the document]
Comment (jeroenbourgois) [reply test/test_helper.exs]: Done

File: lib/myxql/protocol/values.ex [hunk 5]
```diff
@@ -222,7 +222,18 @@ defmodule MyXQL.Protocol.Values do
...
+  def encode_binary_value(%Geo.PointZM{} = geo), do: encode_geometry(geo)
   def encode_binary_value(%Geo.MultiPoint{} = geo), do: encode_geometry(geo)
+  def encode_binary_value(%Geo.LineString{} = geo), do: encode_geometry(geo)
+  def encode_binary_value(%Geo.LineStringZ{} = geo), do: encode_geometry(geo)
+  def encode_binary_value(%Geo.MultiLineString{} = geo), do: encode_geometry(geo)
+  def encode_binary_value(%Geo.MultiLineStringZ{} = geo), do: encode_geometry(geo)
+  def encode_binary_value(%Geo.Polygon{} = geo), do: encode_geometry(geo)
+  def encode_binary_value(%Geo.PolygonZ{} = geo), do: encode_geometry(geo)
+  def encode_binary_value(%Geo.MultiPolygon{} = geo), do: encode_geometry(geo)
+  def encode_binary_value(%Geo.MultiPolygonZ{} = geo), do: encode_geometry(geo)
```
[the remark this answers is not in the document]
Comment (jeroenbourgois) [reply lib/myxql/protocol/values.ex]: There is an issue here, seems that MySQL does not support the Z and M variants of the types... Point is just Point, there is not PointZ.  So I added them but commented them out for now.

File: test/myxql/protocol/values_test.exs [hunk 6]
```diff
@@ -283,6 +283,12 @@ defmodule MyXQL.Protocol.ValueTest do
           coordinates: [{1.0, 1.0}, {2.0, 2.0}]
         })
       end
+
+      test "POLYGON", c do
+        assert_roundtrip(c, "my_polygon", %Geo.Polygon{
+          coordinates: [[{30, 10}, {40, 40}, {20, 40}, {10, 20}, {30, 10}]]
+        })
+      end
```
[the remark this answers is not in the document]
Comment (jeroenbourgois) [reply test/myxql/protocol/values_test.exs]: Still working on this one.

**Comment by jeroenbourgois** [comment]
> @jeroenbourgois oh, I didn't notice you've opened this branch against mine! If it's not too much trouble, would you mind changing base branch to `master`? (or just opening a new PR if that's easier)
> 
> Oh, and if you want to do the honours and remove https://github.com/elixir-ecto/myxql/blame/master/MARIAEX_COMPATIBILITY.md#L41 - by all means :D Thanks again for your help with this!

Would I ?! :D I'll keep this as last.

File: lib/myxql/protocol/values.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (wojtekmach) [reply lib/myxql/protocol/values.ex]: OK, thanks for double-checking this.

In that case, let's not leave commented out code but just remove it.

I found ["Internal Geometry Storage Format"](https://dev.mysql.com/doc/refman/8.0/en/gis-data-formats.html) section in MySQL manual which explains a couple things:

> MySQL stores geometry values using 4 bytes to indicate the SRID followed by the WKB representation of the value.

So yeah, the first 4 byte are indeed for SRID and MySQL in fact does not implement EWKB, but just WKB. This means that even though `Geo.WKB` encodes/decodes SRID we won't be able to use that. I think we'd have to something along the lines of:

```elixir
  defp encode_geometry(geom) do
    srid = geom.srid || 0
    binary = %{geo | srid: nil} |> Geo.WKB.encode!(:ndr) |> Base.decode16!()
    {:mysql_type_geometry, encode_string_lenenc(<<srid::uint4, binary::binary>>)}
  end
```

and something similar for decoding.

> MySQL supports geometry types of Point, LineString, Polygon, MultiPoint, MultiLineString, MultiPolygon, and GeometryCollection. Other geometry types are not supported.

File: test/myxql/protocol/values_test.exs [hunk 7]
```diff
@@ -274,15 +274,119 @@ defmodule MyXQL.Protocol.ValueTest do
...
+      #       [[{40, 40, 40}, {20, 45, 60}, {45, 30, 15}, {40, 40, 40}]],
+      #       [
+      #         [{20, 35, 48}, {10, 30, 50}, {10, 10, 10}, {30, 5, 18}, {45, 20, 10}, {20, 35, 48}],
+      #         [{30, 20, 10}, {20, 15, 10}, {20, 25, 4}, {30, 20, 10}]
+      #       ]
+      #     ]
+      #   })
+      # end
+
+      if @protocol == :text do
```
Comment (wojtekmach) [inline test/myxql/protocol/values_test.exs]: I tried replying over email and doesn't look like GitHub UI shows my response so trying again here:

> We’re missing `def encode_binary_value(%Geo.GeometryCollection{} = geo), do: encode_geometry(geo)`, after that we should be good to go.

The error you were getting about Jason is we attempt to encode lists and maps to the json column, but because structs are maps that clause got invoked.

File: test/myxql/protocol/values_test.exs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeroenbourgois) [reply test/myxql/protocol/values_test.exs]: Done!

File: lib/myxql/protocol/values.ex (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeroenbourgois) [reply lib/myxql/protocol/values.ex]: Good find about the srid, glad this is clear now. I removed all commented stuff.

**Comment by jeroenbourgois** [comment]
@wojtekmach too bad, all but the `mariadb` CI builds pass... :(

**Comment by wojtekmach** [comment]
@jeroenbourgois there're only 2 things left

1. please add me to your fork as a collaborator so that I can push to your branch, I have a couple minor fixes ready to apply

2. please change the base branch of this PR to master

and that's it!

**Comment by jeroenbourgois** [comment]
@wojtekmach done, and done!

**Comment by wojtekmach** [comment]
Thanks to everyone who helped with it!
